### ISSUE TITLE
Invalid AssetGraph after commenting/uncommenting re-export

### ISSUE THREAD

**Issue by yury-yufimov** [dart-lang/sdk#38496]
Versions:
Dart VM version: 2.5.0 (Fri Sep 6 20:39:27 2019 +0200) on "macos_x64"
angular: 5.3.1
build_modules: 2.5.0
build_runner: 1.7.0
build_web_compilers: 2.4.0
webdev: 2.5.0
build_runner_core: 4.0.0

We've encountered following error while performing refactoring: 
when you make a mistake in imports, obviously, build fails. 
But in some cases build can't succeed even after you return code to initial state!

In particular, this happens when you have several entry points in library you depend on, and that entry points share same (transitive) imported component.

There is test project:
https://github.com/yury-yufimov/asset_graph_issue/blob/master/asset_graph_test/README.md

for reproduce you can
1. start `pub run webdev serve` in asset_graph_test
2. comment re-export in `input_comps/lib/text_input.dart`
3. Uncomment it

Error:
```
[SEVERE]build_web_compilers:ddc on package:input_comps/select.template.ddc.module: Error compiling dartdevc module:input_comps|lib/select.template.ddc.js

We're sorry, you've found a bug in our compiler.
You can report this bug at:
    https://github.com/dart-lang/sdk/issues/labels/web-dev-compiler
Please include the information below in your report, along with
any other information that may help us track it down. Thanks!
-------------------- %< --------------------
    dartdevc -k arguments: --dart-sdk-summary=/usr/local/Cellar/dart/2.5.0/libexec/lib/_internal/ddc_sdk.dill --modules=amd --no-summarize -o packages/input_comps/select.template.ddc.js --source-map --summary=packages/meta/meta.ddc.dill=packages/meta/meta --summary=packages/collection/src/utils.ddc.dill=packages/collection/src/utils --summary=packages/collection/src/algorithms.ddc.dill=packages/collection/src/algorithms --summary=packages/collection/src/iterable_zip.ddc.dill=packages/collection/src/iterable_zip --summary=packages/collection/src/comparators.ddc.dill=packages/collection/src/comparators --summary=packages/collection/src/priority_queue.ddc.dill=packages/collection/src/priority_queue --summary=packages/collection/src/canonicalized_map.ddc.dill=packages/collection/src/canonicalized_map --summary=packages/collection/collection.ddc.dill=packages/collection/collection --summary=packages/logging/logging.ddc.dill=packages/logging/logging --summary=packages/path/path.ddc.dill=packages/path/path --summary=packages/stack_trace/src/chain.ddc.dill=packages/stack_trace/src/chain --summary=packages/meta/dart2js.ddc.dill=packages/meta/dart2js --summary=packages/angular/src/core/change_detection/change_detection.ddc.dill=packages/angular/src/core/change_detection/change_detection --summary=packages/angular/src/core/application_tokens.ddc.dill=packages/angular/src/core/application_tokens --summary=packages/angular/core.ddc.dill=packages/angular/core --summary=packages/angular/di.ddc.dill=packages/angular/di --summary=packages/angular/src/core/change_detection/change_detection.template.ddc.dill=packages/angular/src/core/change_detection/change_detection.template --summary=packages/angular/di.template.ddc.dill=packages/angular/di.template --summary=packages/intl/date_symbols.ddc.dill=packages/intl/date_symbols --summary=packages/intl/number_symbols.ddc.dill=packages/intl/number_symbols --summary=packages/intl/number_symbols_data.ddc.dill=packages/intl/number_symbols_data --summary=packages/intl/src/plural_rules.ddc.dill=packages/intl/src/plural_rules --summary=packages/intl/intl.ddc.dill=packages/intl/intl --summary=packages/angular/src/core/application_tokens.template.ddc.dill=packages/angular/src/core/application_tokens.template --summary=packages/angular/core.template.ddc.dill=packages/angular/core.template --summary=packages/js/js_util.ddc.dill=packages/js/js_util --summary=packages/js/js.ddc.dill=packages/js/js --summary=packages/angular/src/bootstrap/modules.ddc.dill=packages/angular/src/bootstrap/modules --summary=packages/angular/angular.ddc.dill=packages/angular/angular --summary=packages/input_comps/src/text_input/text_input.ddc.dill=packages/input_comps/src/text_input/text_input --summary=packages/angular/src/bootstrap/run.template.ddc.dill=packages/angular/src/bootstrap/run.template --summary=packages/angular/angular.template.ddc.dill=packages/angular/angular.template --summary=packages/input_comps/src/text_input/text_input.template.ddc.dill=packages/input_comps/src/text_input/text_input.template --summary=packages/input_comps/src/select/select.ddc.dill=packages/input_comps/src/select/select --summary=packages/input_comps/select.ddc.dill=packages/input_comps/select --packages=file:///var/folders/rz/w8mxl36j3zd2m5dk8rz49nfn3zj2lp/T/kernel_builder_F5j95w/.packages --module-name=packages/input_comps/select.template --multi-root-scheme=org-dartlang-app --multi-root=. --track-widget-creation --inline-source-map --libraries-file=file:///usr/local/Cellar/dart/2.5.0/libexec/lib/libraries.json package:input_comps/select.template.dart package:input_comps/src/select/select.template.dart
    dart --version: 2.5.0 (Fri Sep 6 20:39:27 2019 +0200) on "macos_x64"

Bad state: Could not find module name for library "#lib1" from component "null".
#0      ProgramCompiler.libraryToModule (package:dev_compiler/src/kernel/compiler.dart:392:7)
dart-lang/build#1      SharedCompiler.emitImportsAndExtensionSymbols (package:dev_compiler/src/compiler/shared_compiler.dart:449:27)
dart-lang/build#2      ProgramCompiler.emitModule (package:dev_compiler/src/kernel/compiler.dart:334:5)
dart-lang/build#3      _compile (package:dev_compiler/src/kernel/command.dart:366:27)
<asynchronous suspension>
dart-lang/build#4      compile (package:dev_compiler/src/kernel/command.dart:42:18)
<asynchronous suspension>
dart-lang/build#5      compile (package:dev_compiler/src/compiler/shared_command.dart:394:12)
dart-lang/build#6      _CompilerWorker.performRequest.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/dev_compiler/bin/dartdevc.dart:64:13)
dart-lang/build#7      _rootRun (dart:async/zone.dart:1124:13)
dart-lang/build#8      _CustomZone.run (dart:async/zone.dart:1021:19)
dart-lang/build#9      _runZoned (dart:async/zone.dart:1516:10)
dart-lang/build#10     runZoned (dart:async/zone.dart:1463:12)
dart-lang/build#11     _CompilerWorker.performRequest (file:///b/s/w/ir/cache/builder/sdk/pkg/dev_compiler/bin/dartdevc.dart:62:24)
<asynchronous suspension>
dart-lang/build#12     AsyncWorkerLoop.run.<anonymous closure> (package:bazel_worker/src/worker/async_worker_loop.dart:33:41)
dart-lang/build#13     _rootRun (dart:async/zone.dart:1124:13)
dart-lang/build#14     _CustomZone.run (dart:async/zone.dart:1021:19)
dart-lang/build#15     _runZoned (dart:async/zone.dart:1516:10)
dart-lang/build#16     runZoned (dart:async/zone.dart:1463:12)
dart-lang/build#17     AsyncWorkerLoop.run (package:bazel_worker/src/worker/async_worker_loop.dart:33:26)
<asynchronous suspension>
dart-lang/build#18     main (file:///b/s/w/ir/cache/builder/sdk/pkg/dev_compiler/bin/dartdevc.dart:28:57)
<asynchronous suspension>
dart-lang/build#19     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:301:32)
dart-lang/build#20     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

**Comment by jensjoha** [comment]
@yury-yufimov thank you for the bug report!

@jakemac53 what happens is this:

- The build system packs files into modules.
- Those modules are compiled into dill
- Those dills are used in DDC
- We keep old loaded dills in memory to avoid redoing the same work again and again
- In this instance DDC is first given a module, `text_input.ddc.dill` which contains (among others probably) `package:input_comps/src/text_input/text_input.dart`.
- In a new compilation it is given `select.ddc.dill` which contains the same library (`package:input_comps/src/text_input/text_input.dart`). The library is logically overwritten - but the module it was loaded from before (indexed under `text_input.ddc.dill`) still contains is. Because the library was loaded from a different dill with a different path it is kept in memory.
- In a new compilation we are again asked to load an unchanged `text_input.ddc.dill` (containing `package:input_comps/src/text_input/text_input.dart`) --- which we have in memory, and since it wasn't changed we don't reload it. The library was however just "logically overwritten" so whatever other libraries we have in memory that points to that library, points to the one we overwrote it with above.
- We now have a "clash of worlds" where the same library lives and is active in the same compilation.

There are two solutions to this (that I see):

- Either the build system should not arbitrarily put libraries into modules (i.e. not allow libraries to jump back-and-forth between modules), or
- We detect it and evict old cached dills when loading something new that was also contained in old loads. Somewhat similar as the suggestion I had to a similar problem in bazel here: https://dart-review.googlesource.com/c/sdk/+/113023.

I suggest number two, and will start doing a prototype of sorts of that now.

**Comment by jakemac53** [comment]
Thanks! Yes number 2 sounds like the best option.

There is no way the build system can avoid changing what libraries live in what modules due to the circular dependencies restriction. Any change to directives could introduce a cycle forcing us to re-organize things. In practice we re-organize more often than that most likely, but always in response to either new files being added or directives being modified.

**Comment by jensjoha** [comment]
Fixed in dccbb236a947c7b3d92b2ff5f176e35ffcfd461c.
